fem/integ/bilininteg_diffusion_kernels.hpp

```diff
@@ -9,28 +9,29 @@
 // terms of the BSD-3 license. We welcome feedback and contributions, see file
 // CONTRIBUTING.md for details.
 
-#include "../general/forall.hpp"
-#include "bilininteg.hpp"
-#include "gridfunc.hpp"
-#include "qfunction.hpp"
-#include "ceed/integrators/diffusion/diffusion.hpp"
+#ifndef MFEM_BILININTEG_DIFFUSION_KERNELS_HPP
+#define MFEM_BILININTEG_DIFFUSION_KERNELS_HPP
 
-using namespace std;
+#include "../../config/config.hpp"
+#include "../../general/forall.hpp"
+#include "../../linalg/dtensor.hpp"
 
 namespace mfem
 {
 
-// PA Diffusion Integrator
+namespace internal
+{
 
 // OCCA 2D Assemble kernel
 #ifdef MFEM_USE_OCCA
-static void OccaPADiffusionSetup2D(const int D1D,
-                                   const int Q1D,
-                                   const int NE,
-                                   const Array<double> &W,
-                                   const Vector &J,
-                                   const Vector &C,
-                                   Vector &op)
+MFEM_HOST_DEVICE inline
```

Comment:
I think `inline` is required otherwise we're going to get multiple definition errors (function definition outside of a class body in a header file is going to be defined once per translation unit). As for `MFEM_HOST_DEVICE`, you are right I think since the kernels are launched from inside these functions.

Reply:
As it stands, the reorganization created the `*_kernels.hpp` files to include all the free functions required for integrators (many of which are shared between integrators).

These new files could be separated into `*_kernels.hpp` and `*_kernels.cpp` to save a few `inline`s, but the impact will be minimal since most are template functions and I don't think we are trying to introduce any explicit instantiation. So, I don't really see much value in splitting and keeping them header only seems like a good approach to me.

If alternatively you are suggesting moving some of the functions from the `*_kernels.hpp` files back into the `bilininteg_*.cpp` files, I think this starts to break the separation between integrator class member functions (`AssemblePA`, `AddMultPA`, ...) and the free (often shared) helper functions which launch kernels.

Reply:
As a start, see [9c0eae5](https://github.com/mfem/mfem/pull/3526/commits/9c0eae5e3e43f3f80d34ac8cbdfa9ab56b6215f2) for the `MFEM_HOST_DEVICE` fix.

Reply:
OK, sounds fine to me.

Are you able to build with CUDA? I'm getting warnings like
```
cc1plus: warning: 'void* __builtin_memcpy(void*, const void*, long unsigned int)' writing between 8 and 17179869176 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
```

Reply:
I don't have a test environment set up currently but could do that if needed. Is there any more context to the warnings to try to debug?